### PR TITLE
Constrain Order Parameters for Safety

### PR DESCRIPTION
**Pull request by zachanon**
No order should be placed for a non-positive amount of interest. This PR includes a check before placing an order that bond tickets > underlying tokens.

### CODE REVIEW

File: libraries/ts/bonds/src/bondMarket.ts [hunk 1]
```diff
@@ -176,7 +176,7 @@ export class BondMarket {
   ): Promise<TransactionInstruction> {
     // TODO: rethink amounts here, current is placeholder
     const params: OrderParams = {
-      maxBondTicketQty: amount,
+      maxBondTicketQty: amount.add(new BN(1)),
```
[the remark this answers is not in the document]
Comment (zachanon) [reply libraries/ts/bonds/src/bondMarket.ts:179]: its a bandaid until the fix to the AAOB that gives more detailed return values allows us to properly calculate collateral. Without the +1 this instruction fails the check that interest is positive nonzero

File: programs/bonds/src/errors.rs [hunk 2]
```diff
@@ -22,6 +22,8 @@ pub enum BondsError {
     InsufficientSeeds,
     #[msg("the wrong event type was unwrapped\nthis condition should be impossible, and does not result from invalid input")]
     InvalidEvent,
+    #[msg("interest for an order should always be positive")]
```
[the remark this answers is not in the document]
Comment (zachanon) [reply programs/bonds/src/errors.rs:25]: Noted. For current state, alphabetical is fine. But after release, should use that practice.

File: programs/bonds/src/orderbook/state.rs [hunk 3]
```diff
@@ -89,6 +89,12 @@ impl<'info> OrderbookMut<'info> {
             post_allowed,
             auto_stake: _,
         } = params;
+
+        // No orders should have an interest rate <=0
+        if max_underlying_token_qty >= max_bond_ticket_qty {
```
Comment (dnut) [inline programs/bonds/src/orderbook/state.rs:94]: The ratio of max quantities is not the same as the price. For example you can set max_underlying_token_qty to u64::MAX and max_bond_ticket_qty to 20 with a price of 0.9. That implies a positive interest rate of around 11% that will only be considered filled once 20 bond tickets have been traded.

I would use the `limit_price` field so you can test the actual interest rate.

File: tests/hosted/tests/bonds.rs [hunk 4]
```diff
@@ -237,6 +237,13 @@ async fn _full_workflow<P: Proxy + GenerateProxy>(manager: Arc<BondsTestManager>
     #[cfg(not(feature = "localnet"))]
     manager.consume_events().await?;
 
+    // Cannot make a bad borrow order
+    let bad_params = OrderParams {
+        max_bond_ticket_qty: 0,
```
Comment (dnut) [inline tests/hosted/tests/bonds.rs:242]: The same comment I made for the other order also applies here.

This approach has the additional downside that it won't allow users to specify a borrow order in terms of bond ticket quantity. They are only allowed to do so in terms of underlying. Granted, underlying is probably what they're usually thinking about. But there may be some case like for example, they know exactly how much money they will receive in 6 months, so they just want to borrow as much money as possible today that can definitely be repaid with the income they are certain they will receive in six months.

**Review by dnut — CHANGES_REQUESTED** [review]
If you want to constrain the interest rate, constrain the price. Constraining max quantities will only constrain the order size, or traded volume, not the interest rate.

**Comment by zachanon** [comment]
closing this PR to make a fresh branch

**Pull request closed without merging**